src/test/txvalidation_tests.cpp

```diff
@@ -59,5 +61,571 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChainSetup)
+    testTx(this, MakeTransactionRef(tokenIssueTx), true);
+
+    //token transfer TYPE=3
+    CScript scriptPubKey1 = CScript() << colorid.toVector() << OP_COLOR << OP_DUP << OP_HASH160 << ToByteVector(pubkeyHash1) << OP_EQUALVERIFY << OP_CHECKSIG;
+
+    CScript scriptPubKey2 = CScript() << colorid.toVector() << OP_COLOR << OP_DUP << OP_HASH160 << ToByteVector(pubkeyHash2) << OP_EQUALVERIFY << OP_CHECKSIG;
+    CMutableTransaction tokenTransferTx;
+
+    tokenTransferTx.nFeatures = 1;
+    tokenTransferTx.vin.resize(1);
+    tokenTransferTx.vout.resize(1);
+    tokenTransferTx.vin[0].prevout.hashMalFix = tokenIssueTx.GetHashMalFix();
+    tokenTransferTx.vin[0].prevout.n = 0;
+    tokenTransferTx.vout[0].nValue = 1;
+    tokenTransferTx.vout[0].scriptPubKey = scriptPubKey1;
+
+    Sign(vchSig, key0, tokenIssueTx.vout[0].scriptPubKey, tokenIssueTx, 0, tokenTransferTx, 0);
+    tokenTransferTx.vin[0].scriptSig = CScript() << vchSig << vchPubKey0;
+
+    testTx(this, MakeTransactionRef(tokenTransferTx), false, "bad-txns-token-without-fee");
+
+    tokenTransferTx.vin.resize(2);
+    tokenTransferTx.vin[1].prevout.hashMalFix = m_coinbase_txns[3]->GetHashMalFix();
+    tokenTransferTx.vin[1].prevout.n = 0;
+    tokenTransferTx.vout.resize(2);
+    tokenTransferTx.vout[1].nValue = 1;
+    tokenTransferTx.vout[1].scriptPubKey = scriptPubKey2;
+
+    Sign(vchSig, key0, tokenIssueTx.vout[0].scriptPubKey, tokenIssueTx, 0, tokenTransferTx, 0);
+    tokenTransferTx.vin[0].scriptSig = CScript() << vchSig << vchPubKey0;
+    CMutableTransaction coinbaseIn2(*m_coinbase_txns[3]);
+    Sign(vchSig, coinbaseKey, m_coinbase_txns[3]->vout[0].scriptPubKey, coinbaseIn2, 1, tokenTransferTx, 0);
+    tokenTransferTx.vin[1].scriptSig = CScript() << vchSig;
+
+    testTx(this, MakeTransactionRef(tokenTransferTx), false, "bad-txns-token-balance");
+
+    tokenTransferTx.vout[0].nValue = 1;
+    tokenTransferTx.vout[1].nValue = 0;
+
+    Sign(vchSig, key0, tokenIssueTx.vout[0].scriptPubKey, tokenIssueTx, 0, tokenTransferTx, 0);
+    tokenTransferTx.vin[0].scriptSig = CScript() << vchSig << vchPubKey0;
+    Sign(vchSig, coinbaseKey, m_coinbase_txns[3]->vout[0].scriptPubKey, coinbaseIn2, 1, tokenTransferTx, 0);
+    tokenTransferTx.vin[1].scriptSig = CScript() << vchSig;
+
+    dustRelayFee = CFeeRate(1);
+    testTx(this, MakeTransactionRef(tokenTransferTx), false, "dust");
+
+    dustRelayFee = CFeeRate(0);
+    testTx(this, MakeTransactionRef(tokenTransferTx), false, "invalid-colorid");
+
+    tokenTransferTx.vout.resize(1);
+    tokenTransferTx.vout[0].nValue = 1;
+
+    Sign(vchSig, key0, tokenIssueTx.vout[0].scriptPubKey, tokenIssueTx, 0, tokenTransferTx, 0);
+    tokenTransferTx.vin[0].scriptSig = CScript() << vchSig << vchPubKey0;
+    Sign(vchSig, coinbaseKey, m_coinbase_txns[3]->vout[0].scriptPubKey, coinbaseIn2, 1, tokenTransferTx, 0);
+    tokenTransferTx.vin[1].scriptSig = CScript() << vchSig;
+
+    testTx(this, MakeTransactionRef(tokenTransferTx), true);
+}
 
```

Comment:
add tests for the following cases?
- Transfer of multiple tokens
- Issue and transfer tokens at the same tx